force-app/main/default/classes/SOQL.cls

```diff
@@ -344,18 +419,39 @@ public with sharing class SOQL {
      * Example: setPrimaryContact() should set the 'Primary_Contact__c' field for every Opportunity in the org
      */
     public static void setPrimaryContact() {
-        //Use the getOpportunityDetailsForClosedWon() method to get the Opportunities
-
-        // Create a set of Account Ids from the Opportunities
-
-        // Query the Contacts
-
-        // Loop through each Contact and add the Account Id and Contact to the map
-        
-        // Loop through each Opportunity
-
-                // Update the Opportunity primary contact from accountToContactMap
-
-        // Update the Opportunities
+        // Step 1: Use the getOpportunityDetailsForClosedWon() method to get the Opportunities
+        List<Opportunity> closedWonOpportunities = getOpportunityDetailsForClosedWon();
+    
+        // Step 2: Create a set of Account Ids from the Opportunities
+        Set<Id> accountIds = new Set<Id>();
+        for (Opportunity opp : closedWonOpportunities) {
+            accountIds.add(opp.AccountId);
+        }
+    
+        // Step 3: Query the Contacts for these Account Ids
+        List<Contact> contacts = [SELECT Id, AccountId FROM Contact WHERE AccountId IN :accountIds ORDER BY LastName ASC LIMIT 1];
+    
+        // Step 4: Map Account Id to the first Contact for that Account
+        Map<Id, Contact> accountToContactMap = new Map<Id, Contact>();
+        for (Contact con : contacts) {
+            if (!accountToContactMap.containsKey(con.AccountId)) {
+                accountToContactMap.put(con.AccountId, con);
+            }
+        }
+    
+        // Step 5: Loop through each Opportunity and update its Primary Contact
+        List<Opportunity> opportunitiesToUpdate = new List<Opportunity>();
+        for (Opportunity opp : closedWonOpportunities) {
+            if (accountToContactMap.containsKey(opp.AccountId)) {
+                opp.Primary_Contact__c = accountToContactMap.get(opp.AccountId).Id;
+                opportunitiesToUpdate.add(opp);
```

Comment:
Try to remove the opportunitiesToUpdate list and use closeWonOpportunities  to do the update